src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/StringBuilderAppendShouldNotTakeSubstring.Fixer.cs

```diff
@@ -0,0 +1,191 @@
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Analyzer.Utilities;
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, LanguageNames.VisualBasic), Shared]
+    public sealed class StringBuilderAppendShouldNotTakeSubstringFixer : CodeFixProvider
+    {
+        public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray
+            .Create(
+                StringBuilderAppendShouldNotTakeSubstring.RuleIdOneParameterId,
+                StringBuilderAppendShouldNotTakeSubstring.RuleIdTwoParameterId);
+
+        public sealed override FixAllProvider GetFixAllProvider()
+        {
+            return WellKnownFixAllProviders.BatchFixer;
+        }
+
+        public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            foreach (var diagnostic in context.Diagnostics)
+            {
+                if (diagnostic.Id == StringBuilderAppendShouldNotTakeSubstring.RuleIdOneParameterId)
+                {
+                    context
+                        .RegisterCodeFix(
+                            new MyCodeAction(
+                                MicrosoftApiDesignGuidelinesAnalyzersResources.StringBuilderShouldUseSubstringOverloadFix,
+                                ct => FixCodeOneParameter(context.Document, context.Span, ct)),
+                            diagnostic);
+                }
+                else
+                {
+                    context
+                        .RegisterCodeFix(
+                            new MyCodeAction(
+                                MicrosoftApiDesignGuidelinesAnalyzersResources.StringBuilderShouldUseSubstringOverloadFix,
+                                ct => FixCodeTwoParameters(context.Document, context.Span, ct)),
+                            diagnostic);
+                }
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private static async Task<Document> FixCodeOneParameter(
+            Document document,
+            TextSpan span,
+            CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var nodeToFix = root.FindNode(span, getInnermostNodeForTie: true);
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var typedNodeToFix =  semanticModel.GetOperation(nodeToFix) as IInvocationOperation;
```

Comment:
why s this an 'as'?

Reply:
tehre is nothing about hte analyzer path that validates this will succeed.

Reply:
Yes, please add a null check that `typedNodeToFix != null` and bail out if it is null.

---
In reply to: [226061455](https://github.com/dotnet/roslyn-analyzers/pull/1854#discussion_r226061455) [](ancestors = 226061455)